### ISSUE TITLE
The metaclass of a derived class must be a subclass of the metaclasses of all its base classes

### ISSUE THREAD

**Issue by ethanabrooks** [microsoft/pyright#1583]
For the following code:
```python
import haiku as hk


class DenseTorso(hk.Module):
    pass
```
`pyright` raises:
```
❯ pyright --lib main.py
No configuration file found.
stubPath /Users/ethanbrooks/pyright-issue/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
/Users/ethanbrooks/pyright-issue/main.py
  5:7 - error: The metaclass of a derived class must be a subclass of the metaclasses of all its base classes (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos 
Completed in 0.587sec
```

I could not reproduce this with simple examples that do not involve an import, so this probably has something to do with how pyright handles imports. I thought that the metaclass error had to do with multiple inheritance situations (as dicussed [here](http://www.phyast.pitt.edu/~micheles/python/metatype.html)). The relevant source code for the `hk.Module` is

```python
class Module(object, metaclass=ModuleMetaclass):
    ...
```
As you can see, a custom metaclass but no multiple inheritance.

**Comment by erictraut** [comment]
I'm not able to repro the problem as reported. I presume that `haiku` is the "dm-haiku" library? I installed it from pip. Do you have any type stubs installed for this library? I couldn't find any. I tried with both "useLibraryCodeForTypes" enabled and disabled.

Are you using the latest version of pyright (1.1.118)?

**Comment by ethanabrooks** [comment]
I was able to reproduce with this Dockerfile:
```Dockerfile
FROM python:3.8-buster

# install pyright
USER root
WORKDIR /home/app
RUN apt-get update -q && \
  apt-get -y install curl gnupg 
RUN curl -sL https://deb.nodesource.com/setup_15.x | bash -
RUN apt-get -y install nodejs
RUN npm install -g pyright

# install haiku
RUN pip install "git+https://github.com/deepmind/dm-haiku.git@b2a7582c3721d07493caaf20817595c27c6e0e3a"

# create main.py
RUN echo '\
import haiku as hk \n\
class DenseTorso(hk.Module): \n\
    pass \n'\
     >> main.py

CMD ["pyright", "--lib"]
```

**Comment by erictraut** [comment]
OK, thanks for the additional steps. This doesn't occur with the latest published version of dm-haiku. It's apparently related to a recent change made in that code base. Upon initial inspection, it looks like it's because `ModuleMetaclass` now subclasses from `type(Protocol)`, which is a very odd construct — one that pyright doesn't understand. I'll need to do some research to understand why the authors of this library did this and what they intended by it.

**Comment by erictraut** [comment]
I've implemented a workaround that eliminates the false positive error message. The fix will be in the next release.

**Comment by ethanabrooks** [comment]
Hi Eric. Thanks for your help! I was still able to reproduce the issue with the latest Haiku commit:
```python
Sending build context to Docker daemon  124.9kB
Step 1/9 : FROM python:3.8-buster
 ---> 9b9126f2a963
Step 2/9 : WORKDIR /home
 ---> Using cache
 ---> 3dc7bd45429c
Step 3/9 : RUN apt-get update -q &&   apt-get -y install curl gnupg
 ---> Using cache
 ---> 6707176f6a39
Step 4/9 : RUN curl -sL https://deb.nodesource.com/setup_15.x | bash -
 ---> Using cache
 ---> 4e0d6b242e34
Step 5/9 : RUN apt-get -y install nodejs
 ---> Using cache
 ---> b2e28664cd87
Step 6/9 : RUN npm install -g pyright
 ---> Using cache
 ---> 5e7c2c1d3add
Step 7/9 : RUN pip install "git+https://github.com/deepmind/dm-haiku.git@b2a7582c3721d07493caaf20817595c27c6e0e3a"
 ---> Using cache
 ---> 2e088dff8bd4
Step 8/9 : RUN echo 'import haiku as hk \nclass DenseTorso(hk.Module): \n    pass \n'     >> main.py
 ---> Using cache
 ---> 53c74644d748
Step 9/9 : CMD ["pyright", "--lib"]
 ---> Using cache
 ---> f676c6cea292
Successfully built f676c6cea292
Successfully tagged pyright-issue:latest
```

Can I either build from source or install a nightly release to get your fix?

**Comment by erictraut** [comment]
I typically publish two or three times a week depending on when I have enough fixes ready to justify a release.

Of course, you're free to build from source if you want to do so.

**Comment by erictraut** [comment]
This is included in Pyright 1.1.119, which I just published. It will also be included in the next version of Pylance.

**Comment by ethanabrooks** [comment]
Hi, one final question: would you happen to know how to point VS code to a local install of pyright? Thanks.

**Comment by erictraut** [comment]
I'm not sure what you mean by "local install of pyright"? Are you building it from scratch yourself? If so, are you doing it in debug mode, or are you building a VSIX package? Instructions for both can be found [here](https://github.com/microsoft/pyright/blob/master/docs/build-debug.md).

**Comment by ethanabrooks** [comment]
I see. Thanks!
